src/crypto/ed448_suite.c

```diff
@@ -0,0 +1,205 @@
+// Include
+#include <decaf/ed448.h>
+#include <decaf/shake.h>
+
+
+#include "ed448_suite.h"
+#include <stdio.h>
+
+wickr_buffer_t *ed448_sig_derive_public_key(const wickr_buffer_t *private_key_data)
+{
+    if (!private_key_data || private_key_data->length != EDDSA_448_PRIVATE_KEY_LENGTH)
+        return NULL;
+    
+    wickr_buffer_t *public_key_data = wickr_buffer_create_empty_zero(EDDSA_448_PUBLIC_KEY_LENGTH);
+
+    if (!public_key_data)
+        return NULL;
+    
+    decaf_ed448_derive_public_key(public_key_data->bytes, private_key_data->bytes);
+    // This function does not allow for failure
+
+    return public_key_data;    
+}
+
+wickr_ecdsa_result_t *ed448_sig_sign(const wickr_ec_key_t *ec_signing_key,
+                               const wickr_buffer_t *data_to_sign,
+                               wickr_digest_t digest_mode)
+{
+    if (!ec_signing_key)
+        return NULL;
+
+    if (ec_signing_key->curve.identifier != EC_CURVE_ID_ED448_GOLDILOCKS)
+        return NULL;
+
+    if (!data_to_sign)
+        return NULL;
+
+    const wickr_buffer_t *private_data = ec_signing_key->pri_data;
+    const wickr_buffer_t *public_data = ec_signing_key->pub_data;
+
+    if (!private_data || !public_data)
+        return NULL;
+
+    if (private_data->length != EDDSA_448_PRIVATE_KEY_LENGTH ||
+        public_data->length != EDDSA_448_PUBLIC_KEY_LENGTH)
+        return NULL;
+
+    wickr_buffer_t *signature = wickr_buffer_create_empty_zero(EDDSA_448_SIGNATURE_LENGTH);
+    uint8_t prehashed = 0;
+    uint8_t * context = NULL;
+    uint8_t context_length = 0;
+
+    //TODO(Michal): different behaviour based on digest mode
+    decaf_ed448_sign(signature->bytes, private_data->bytes, public_data->bytes, data_to_sign->bytes,
+                     data_to_sign->length, prehashed, context, context_length);
+    // Library function has no failure case
+    
+    wickr_ecdsa_result_t *result = wickr_ecdsa_result_create(ec_signing_key->curve, digest_mode, signature);
```

Comment:
any of my create functions can in theory return NULL, if that is the case, you need to destroy the signature buffer so we don't leak memory

Reply:
Good catch!